engine/swarm/store_sensitive_strings.md

```diff
@@ -0,0 +1,234 @@
+---
+title: Manage sensitive strings (secrets) for Docker services
+description: How to securely store, retrieve, and use sensitive strings (secrets) in Docker services
+keywords: swarm, secrets, credentials, sensitive strings, security, encryption, encryption at rest
+---
+
+## About secrets
+
+In terms of Docker Swarm services, a _secret_ is a string, such as a password,
+credit card number, SSH private key, SSL certificate, or another piece of
+sensitive data that should not be transmitted over a network or stored in a
+Dockerfile or in your application's source code in plain text. Starting in
+Docker 1.13, Docker can encrypt and manage these secrets for you. Consider the
+following examples when you might want to manage sensitive strings using Docker:
+
+- Your service relies on a password-protected database, and you want to protect
+  the database's credentials.
+- You need a central location to administer and manage credentials needed by
+  different services running in your swarm.
+- Your application relies upon a SSH private key, and you want to protect the
+  key.
+- You run an application, such as a web server, which needs to use a SSL
+  certificate, and you want to protect the certificate itself.
+- You use a different SSL certificate in development and production environments
+  but do not want to change the image or store the SSL certificate in the image.
+- Your application uses infrastructure which is metered and needs a credential,
+  such as Amazon AWS, EC2, or S3, and you want to prevent unauthorized use of
+  this infrastructure.
+- You are a developer building an application which processes sensitive data
+  to which you do not need or want direct access. Your administrator may
+  populate the secret key with a mock value while you are developing the
+  application, and change it to use the real value in production.
+
+In these and many other situations, managing sensitive strings using Docker
+secrets makes sense.
+
+## How Docker manages secrets
+
+When you add a secret to the swarm, Docker sends the secret to each manager
+node over a network connection which uses mutual TLS. When a manager receives
+a secret, it encrypts it with an encryption key before writing it to
+the RAFT logs, just as other RAFT log data is encrypted. The encryption key
+itself has already been propagated to the swarm node using mutual TLS.
+
+>**Warning**: RAFT data is encrypted in Docker 1.13 and newer. If some of your
+Swarm nodes run an earlier version, the secrets are stored unecrypted in
```

Comment:
@diogomonica While RAFT messages may be written to disk opaquely, I'm not sure that a manager with 1.12 snapshot structure definition will actually do this. None of our types have opaque field preservation, as far as I can tell. Looks like this was removed in proto3, as described in https://github.com/google/protobuf/issues/272.

> The upgrade path should be: upgrade all the managers to 1.13, and then start using secrets.

I agree and this language is sufficient to protect users against leaks, even if not precisely accurate. We may want to make this stronger and warn users to make sure the _entire_ cluster is on 1.13 before performing cluster mutations.

The only data loss scenario @aaronlehmann and I could work out was if a 1.12 engine was elected leader in a mixed-mode cluster. We may be able to downweight older engines in the master election process to minimize this possibility.

The other scenarios for data loss considered are 1.12 proxying to 1.13 leader, but this is a non-issue, since the JSON API prevents opaque field transfer.